### DISCUSSION TITLE
DataFrame Clean Up

### DISCUSSION THREAD

**Discussion by hanried** [earthlab-education/Earth-Analytics-AY24#348] [Q&A]
What I was doing: In Part 2 of the climate coding challenge, when cleaning up the data frame, I was experimenting with the columns that would show up in the data frame and tried adding “STATION” at the beginning of the double brackets (yellow highlight in screenshot).

Question: Why did I get the “KeyError: “[‘STATION’] not in index” (green highlight)? 

What I thought would happen: I assumed that the original data frame with all three original columns: STATION, PRCP, and TOBS, would've shown up.

<img width="707" alt="image" src="https://github.com/user-attachments/assets/913e4175-03fe-4655-aa95-2ef034181758">

**Comment by eculler** [comment]
Hi @hanried, I'm guessing that this is happening because STATION is not actually a column in your DataFrame. I believe you're correct that STATION was _originally_ in the DataFrame, so this is a little bit tricky. However, I notice that you are writing over the original DataFrame in this cell where you are selecting the columns. SO, what I think happened, is that you:

1. Imported the data (climate_df has a STATION column)
2. Selected only the PRCP and TOBS columns (climate_df now does _not_ have a STATION column)
3. Tried to select STATION, PRCP, and TOBS, which fails because climate_df was written over

You could test this theory by running some code that calls climate_df again to see if the STATION column is still there.

**Reply by eculler** [reply]
This is a small error (you can easily move on with your day by not trying to select the STATION column), but it does a great job of highlighting one of the big challenges with getting started with coding in interactive notebooks like this Jupyter notebook (.ipynb). When I learned to code, we pretty much exclusively wrote **executable scripts** that ran from beginning to end every time. The interactive notebooks are great, because you can run the code in small sections and play around with the results! However, you can **also** run code out of order. To avoid this type of error, I recommend two general approaches:

1. Once you have saved something to a variable name (in this case, climate_df), don't write over that same name. (I know I did this in the example -- it's super easy to forget!) This can lead to problems if you try to run the same code twice. So, in this case you could give it a different name when you select the columns. I like to save the shorter names for my final data, so the code might looks something like:

     ```python
    climate_df_raw = pd.read_csv(climate_url)
    ```
    and then:
    ```python
    climate_df = climate_df_raw[['PRCP', 'TOBS']]
    ```
    Note that we can also get around this problem by putting both lines in the same cell -- that way you also can't run one without the other.

2. When you're coding, make sure to periodically Restart the kernel and run your code from the beginning. I make a note to show this in demo videos -- I often forget by the time I get to the end of the video but it's an important step! This is an important part of reproducibility, as it helps ensure that your code will run for other people if they run it from the top.

**Reply by BioNCphilic** [reply]
Thanks for the explanation of over-writing and how to fix it. I like that you are playing with the code and making predictions @hanried
